qiskit/extensions/hamiltonian_gate.py

```
@@ -0,0 +1,118 @@
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+Gate described by the time evolution of a Hermitian Hamiltonian operator.
+"""
+
+import numpy
+import scipy.linalg
+
+from qiskit.circuit import Gate, QuantumCircuit, QuantumRegister
+from qiskit.quantum_info.operators.predicates import matrix_equal
+from qiskit.quantum_info.operators.predicates import is_hermitian_matrix
+from qiskit.extensions.exceptions import ExtensionError
+
+from .unitary import UnitaryGate
+
+
+class HamiltonianGate(Gate):
+    """Class for representing evolution by a Hermitian Hamiltonian operator as a gate. This gate
+    resolves to a UnitaryGate, which can be decomposed into basis gates if it is 2 qubits or
+    less, or simulated directly in Aer for more qubits. """
+
+    def __init__(self, data, time, label=None):
+        """Create a gate from a hamiltonian operator and evolution time parameter t
+
+        Args:
+            data (matrix or Operator): unitary operator.
+            time (float or complex): time evolution parameter.
+            label (str): unitary name for backend [Default: None].
+
+        Raises:
+            ExtensionError: if input data is not an N-qubit unitary operator.
+        """
+        if hasattr(data, 'to_matrix'):
+            # If input is Gate subclass or some other class object that has
+            # a to_matrix method this will call that method.
+            data = data.to_matrix()
+        elif hasattr(data, 'to_operator'):
+            # If input is a BaseOperator subclass this attempts to convert
+            # the object to an Operator so that we can extract the underlying
+            # numpy matrix from `Operator.data`.
+            data = data.to_operator().data
+        # Convert to numpy array in case not already an array
+        data = numpy.array(data, dtype=complex)
+        # Check input is unitary
+        if not is_hermitian_matrix(data):
```

We should probably throw an exception if time is complex or imaginary too